### PR TITLE
Achievements

### LINKED ISSUE #301: Achievements
# **Description**
**Task:** Achievements.
**Feature:** #269 

Develop an Achievements Tracking Component that constantly monitors the player's current stats and inventory. When player completes certain challenges, a badge is awarded to them. This will help in providing real-time feedback and rewards to players based on their in-game accomplishments. These achievements will be permanently stored with the player.

**Subtasks:**
1. **Design Achievements System:**
   - Define the criteria for various achievements .
   - Determine the corresponding events to be triggered when criteria are met.
   - Plan how achievements will be displayed or communicated to the player.

2. **Implement Achievements Component:**
   - Create the `AchievementsComponent` class that will be attached to the player entity.
   - Add fields to track necessary stats (e.g., weapons collected, enemies killed).
   - Implement methods to update stats based on in-game actions.

3. **Event Listeners and Triggers:**
   - Implement logic to check if achievement criteria are met and trigger appropriate events when they are.

4. **Saving**
   - Save all the badges collected permanently in the game so that every time the game is restarted, the badges collected in previous games are rendered.
.

## **Example (optional)**
- if player is kills ten animals, they will be awarded a badge 
- If player is able to kill the boss without using any items throughout the game, they will be award another badge

# **Dependencies**
- [x] implementation of rooms #8 

 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Design Achievements System (Sep 21)
- [x] Implement Achievements Component (Sep 25)
- [x] Event Listeners and Triggers (Sep 25)
- [x] Saving (Sep 27)
- [x] Testing (Sep 30)
- [x] Documentation (Oct 1)
 
**Completion Deadline:** Oct 2

 
# **Documentation**
- TBA
 
# **Member**
- Manya @manya-k 
- Jas @Jas2310

### PR DESCRIPTION
**Pull request by Jas2310**
# Description
Addition of a new feature that allows players to earn certain "achievements" depending on the tasks undertaken. A UI element of this feature was also implemented, where a small display alerts the player of the achievement won. This feature keeps track of the achievements won by a player, meaning that if the player cannot win the same achievement twice throughout the game, or even if they exit the game and return back in. 

Fixes / Closes #301 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [x] Manual testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            this.table.row().padTop(5f);
+            this.table.add(achievementLabel).padTop(5f);
+            stage.addActor(table);
+            // unrender the label after 1 second of display
+            Timer.schedule(new Timer.Task() {
+                @Override
+                public void run() {
+                    achievementLabel.remove(); // Remove the label from the screen
+                }
+            }, 3);
```
Comment (SIN-Willett) [inline source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java:176]: magic 3, could you make this a constant called "ACHIEVEMENT_DURATION" or something of the like

**Review by SIN-Willett — CHANGES_REQUESTED** [review]

File: source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java (hunk 1 above)
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java]: resolved and pushed !

File: source/core/src/main/com/csse3200/game/components/npc/NPCDeathHandler.java [hunk 2]
```diff
@@ -31,6 +31,7 @@ public class NPCDeathHandler extends Component {
     private boolean isDead = false;
     private Entity target;
     private int npcStrength;
+    private static int deadCount;
```
[the remark this answers is not in the document]
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/components/npc/NPCDeathHandler.java:34]: Will change the implementation in the next sprint

File: source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    public ArrayList<String> energyDrinksCollected;
+    public Table table;
+
+    /**
+     * Constructs an Achievement component to an Entity by reading the achievements
+     * gained through previous games
+     */
+    public PlayerAchievementComponent() {
+        energyDrinksCollected = new ArrayList<>();
+        if (FileLoader.readClass(HashMap.class, "configs/achievements.json",
```
[the remark this answers is not in the document]
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/components/player/PlayerAchievementComponent.java]: resolved!

**Review by dhruvanil55 — APPROVED** [review]
LGTM

**Pull request merged**